src/main/java/software/amazon/awssdk/crt/http/NonBlockingHttpResponsekHandler.java

```diff
@@ -0,0 +1,130 @@
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PipedOutputStream;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Receives calls from Native code and translates native data to/from Java Objects
+ */
+class NonBlockingHttpResponsekHandler implements JniHttpCallbackHandler {
+    private final HttpRequest request;
+    private final CompletableFuture<HttpResponse> responseHandler;
+    private final HttpResponse.Builder responseBuilder = new HttpResponse.Builder();
+    private final PipedOutputStream out;
+    private final AwsCrtPipedInputStream in;
+
+    public NonBlockingHttpResponsekHandler(HttpRequest request, CompletableFuture<HttpResponse> responseHandler) throws IOException {
+        this.request = request;
+        this.responseHandler = responseHandler;
+        out = new PipedOutputStream();
+        AwsCrtInputStreamEvent callback = null; // TODO: wire up callback to update Window sizes on InputStream Read Events
+        in = new AwsCrtPipedInputStream(out, callback);
+    }
+
+    /**
+     * Called from Native when new Http Headers have been received.
+     * Note that this function may be called multiple times as HTTP headers are received.
+     * @param responseStatusCode The HTTP Response Status Code
+     * @param nextHeaders The headers received in the latest IO event.
+     */
+    public void onHeaders(HttpStream stream, int responseStatusCode, HttpHeader[] nextHeaders) {
+        responseBuilder.withStatusCode(responseStatusCode);
+        responseBuilder.withHeaders(nextHeaders);
+    }
+
+    /**
+     * Called from Native once all HTTP Headers are processed.
+     * @param hasBody True if the HTTP Response had a Body, false otherwise.
+     */
+    public void onHeadersDone(HttpStream stream, boolean hasBody) {
+        if (hasBody) {
+            responseBuilder.withBodyStream(in);
+        }
+        responseHandler.complete(responseBuilder.build());
+    }
+
+    /**
+     * Called when new Body bytes have been received.
+     * Note that this function may be called multiple times as headers are received.
+     * Do not keep a reference of the ByteBuffer after this method returns.
+     * @param bodyBytesIn The HTTP Body Bytes received in the last IO Event.
+     * @return The new IO Window Size
+     */
+    public void onResponseBody(HttpStream stream, ByteBuffer bodyBytesIn) {
+        if (!responseHandler.isDone()) {
+            /* May happen if there were no Headers in Response */
+            responseBuilder.withBodyStream(in);
+            responseHandler.complete(responseBuilder.build());
+        }
+
+        try {
+            /* Write the bytes, then update the Native Window Size*/
+            out.write(bodyBytesIn.array());
+            stream.updateWindowSize(in.writeSpaceAvailable());
+        } catch (IllegalStateException|IOException e) {
+            // TODO: What to do?
```

Comment:
Currently, if something goes wrong in a callback you should close the connection from the callback.

The function for closing the connection doesn't currently allow an error-code to be passed, but that's going to  change in the near-future.